README.md
# grunt-mocha-phantomjs

> A simple wrapper to run client-side mocha tests using [mocha-phantomjs](http://metaskills.net/mocha-phantomjs/) core library

[![Build Status](https://travis-ci.org/jdcataldo/grunt-mocha-phantomjs.svg?branch=master)](https://travis-ci.org/jdcataldo/grunt-mocha-phantomjs)
[![Downloads](https://img.shields.io/npm/dm/grunt-mocha-phantomjs.svg)](https://www.npmjs.com/package/grunt-mocha-phantomjs)

## Getting Started
This plugin requires Grunt `~0.4.0`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-mocha-phantomjs --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-mocha-phantomjs');
```

## The "mocha_phantomjs" task

_Run this task with the `grunt mocha_phantomjs` command._

Task targets, files and options may be specified according to the grunt [Configuring tasks](http://gruntjs.com/configuring-tasks) guide.

[PhantomJS][] is installed when installing using NPM.

[PhantomJS]: http://www.phantomjs.org/

### Options

#### reporter
Type: `String`  
Default: `spec`

The reporter that should be used. See [the supported reporters](https://github.com/metaskills/mocha-phantomjs#supported-reporters) for more information.

#### output
Type: `String`  

The file that the task should output the results to. If `output` is specified, the task will always complete and not throw an error code if errors are found. The CI will determine if the build failed or not.

#### failWithOutput
Type: `Boolean`  

Setting `failWithOutput` to true when used with the `output` option will fail a build along with creating a test artifact. This is useful for CIs (CircleCI) that don't fail a build based on the test artifact.

#### silent
Type: `Boolean`  

Setting `silent` to true will prevent the results from being printed using stdout.

#### urls
Type: `Array`  
Default: `[]`

Absolute `http://` or `https://` urls to be passed to PhantomJS. Specified URLs will be merged with any specified `src` files first. Note that urls must be served by a web server, and since this task doesn't contain a web server, one will need to be configured separately. The [grunt-contrib-connect plugin](https://github.com/gruntjs/grunt-contrib-connect) provides a basic web server.

Additional arguments may be passed. See [mocha-phantomjs's](https://github.com/metaskills/mocha-phantomjs#usage) usage.

#### phantomConfig
Type: `Object`
Default: `{}`

Options to be passed directly to phantomjs. Eg:

```js
{
    "--local-storage-path": "my/temp-phantom-files",
    "--local-storage-quota": "20480"
}
```

See `phantomjs -h` for more full options list.

#### config
Type: `Object`  
Default: `{ useColors: true }`

Options to be passed to mocha-phantomjs. See [mocha-phantomjs-core's](https://github.com/nathanboktae/mocha-phantomjs-core#usage) usage.

### Usage examples

#### Basic usage (CI checks for error code)

```js
// Project configuration.
grunt.initConfig({
  mocha_phantomjs: {
    all: ['test/**/*.html']
  }
});
```

#### File output for CI

```js
// Project configuration.
grunt.initConfig({
  mocha_phantomjs: {
    options: {
      reporter: 'xunit',
      output: 'tests/results/result.xml'
    },
    all: ['test/**/*.html']
  }
});
```

#### Passing options to mocha-phantomjs

```js
// Project configuration.
grunt.initConfig({
  mocha_phantomjs: {
    options: {
      reporter: 'xunit',
      output: 'tests/results/result.xml',
      config: {
        useColors: false,
        viewportSize: {
            width: 1024,
            height: 768
        },
        grep: 'pattern'
      }
    },
    all: ['test/**/*.html']
  }
});
```

#### Local server
Include the [grunt-contrib-connect plugin][] to run a local server
[grunt-contrib-connect plugin]: https://github.com/gruntjs/grunt-contrib-connect

```js
// Project configuration.
grunt.initConfig({
  mocha_phantomjs: {
    all: {
      options: {
        urls: [
          'http://localhost:8000/test/foo.html',
          'http://localhost:8000/test/bar.html'
        ]
      }
    }
  },
  connect: {
      server: {
        options: {
          port: 8000,
          base: '.',
        }
      }
    }
});

grunt.registerTask('test', ['connect', 'mocha_phantomjs']);
```

## Release History

* 2016-02-05   v2.0.1   adds a phantomConfig option for setting phantomjs specific settings
* 2015-09-02   v2.0.0   adds a failWithOutput option for failing a build and also providing a test artifact
* 2015-08-09   v1.0.2   Fix path to phantomjs binary on windows
* 2015-08-09   v1.0.1   Fix issue when using --color=false
* 2015-07-16   v1.0.0   *BREAKING* use mocha-phantomjs-core
* 2015-07-16   v0.7.0   Upgrade mocha-phantomjs to 3.6.0
* 2015-07-15   v0.6.2   Lock down phantomjs to match mocha-phantomjs peerDependency
* 2015-02-26   v0.6.1   Add silent option to suppress stdout
* 2014-07-24   v0.6.0   Upgrade mocha-phantomjs to 3.5.0 and drops node 0.8 support
* 2014-05-08   v0.5.0   Upgrade mocha-phantomjs to 3.4.0
* 2014-03-01   v0.4.3   Add lodash and async as dependencies
* 2014-02-18   v0.4.2   Fix to prevent the stream from closing
* 2014-02-18   v0.4.1   Pipe stdout to a file if supplied
* 2014-01-14   v0.4.0   Upgrade mocha-phantomjs to 3.3.0
* 2013-10-31   v0.3.2   Upgrade mocha-phantomjs to 3.2.0
* 2013-10-31   v0.3.1   Added support for --no-color
* 2013-07-05   v0.3.0   Upgrade mocha-phantomjs to 3.1.0
* 2013-04-19   v0.2.8   Fix path to mocha-phantomjs binary on windows when not installed with plugin
* 2013-04-19   v0.2.7   Fixes error thrown when tests pass and no output file specified
* 2013-04-18   v0.2.6   Fix path to mocha-phantomjs binary on windows
* 2013-04-18   v0.2.5   Added an output option to write test results to a file for CI support
* 2013-04-15   v0.2.2   Added exit code to report failed tests with Travis CI
* 2013-04-11   v0.2.1   Added check for existing local install of mocha-phantomjs
* 2013-04-08   v0.2.0   Update phantomjs to 1.9.0 to fix unzipping issue
* 2013-03-27   v0.1.1   Fix to omit urls from being passed to phantomjs
* 2013-03-27   v0.1.0   Initial release supporting all options for mocha-phantomjs


### Notes
This is a very basic implementation of mocha-phantomjs. Failed tests and errors do not bubble up for custom reporting. The idea of this is to be mainly used by a CI and let the CI manage the error reporting.


CHANGELOG.md
## 4.0.0
- adds grunt 1.0 support

## 3.0.0
- upgrades to phantomjs2
- drops node 0.10.x support

## 2.0.1 (2016-03-05)
- adds a phantomConfig option for setting phantomjs specific settings

## 2.0.0 (2015-09-02)
- adds a failWithOutput option for failing a build and also providing a test artifact

## 1.0.2 (2015-08-09)
- Fix path to phantomjs binary on windows

## 1.0.1 (2015-08-09)
- Fix issue when using --color=false

## 1.0.0 (2015-07-16)
- *BREAKING* use mocha-phantomjs-core

## 0.7.0 (2015-07-16)
- Upgrade mocha-phantomjs to 3.6.0

## 0.6.2 (2015-07-15)
- Lock down phantomjs to match mocha-phantomjs peerDependency

## 0.6.1 (2015-02-26)
- Add silent option to suppress stdout

## 0.6.0 (2014-07-24)
- Upgrade mocha-phantomjs to 3.5.0 and drops node 0.8 support

## 0.5.0 (2014-05-08)
- Upgrade mocha-phantomjs to 3.4.0

## 0.4.3 (2014-03-01)
- Add lodash and async as dependencies

## 0.4.2 (2014-02-18)
- Fix to prevent the stream from closing

## 0.4.1 (2014-02-18)
- Pipe stdout to a file if supplied

## 0.4.0 (2014-01-14)
- Upgrade mocha-phantomjs to 3.3.0

## 0.3.2 (2013-10-31)
- Upgrade mocha-phantomjs to 3.2.0

## 0.3.1 (2013-10-31)
- Added support for --no-color

## 0.3.0 (2013-07-05)
- Upgrade mocha-phantomjs to 3.1.0

## 0.2.8 (2013-04-19)
- Fix path to mocha-phantomjs binary on windows when not installed with plugin

## 0.2.7 (2013-04-19)
- Fixes error thrown when tests pass and no output file specified

## 0.2.6 (2013-04-18)
- Fix path to mocha-phantomjs binary on windows

## 0.2.5 (2013-04-18)
- Added an output option to write test results to a file for CI support

## 0.2.2 (2013-04-15)
- Added exit code to report failed tests with Travis CI

## 0.2.1 (2013-04-11)
- Added check for existing local install of mocha-phantomjs

## 0.2.0 (2013-04-08)
- Update phantomjs to 1.9.0 to fix unzipping issue

## 0.1.1 (2013-03-27)
- Fix to omit urls from being passed to phantomjs

## 0.1.0 (2013-03-27)
- Initial release supporting all options for mocha-phantomjs


test/test.js
describe('Test', function () {
  describe('should pass', function () {
    it('should equal 2', function () {
      expect(1 + 1).to.equal(2);
    });

    it('should not equal 3', function () {
      expect(1 + 1).to.not.equal(3);
    });

    it('should equal 3', function () {
      expect(1 + 1 + 1).to.equal(3);
    });
  });
});


tasks/mocha_phantomjs.js
/*
 * grunt-mocha-phantomjs
 * https://github.com/jdcataldo/grunt-mocha-phantomjs
 *
 * Copyright (c) Justin Cataldo <jdcataldo@gmail.com>
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
  var objectAssign = require('object-assign'),
      async        = require('async'),
      path         = require('path'),
      fs           = require('fs');

  function lookup (script, executable) {
    var i = 90,
        absPath;

    for (i = 0; i < module.paths.length; i++) {
      absPath = path.join(module.paths[i], script);
      if (executable && process.platform === 'win32') {
        absPath += '.cmd';
      }
      if (fs.existsSync(absPath)) {
        return absPath;
      }
    }
    grunt.fail.warn('Unable to find ' + script);
  };

  function flatten (arr) {
    var flattened = [];
    return flattened.concat.apply(flattened, arr);
  };

  function objectToArgArray (obj) {
    var key,
        args = [];

    for (key in obj) {
      if (obj.hasOwnProperty(key)) {
        args.push(key + '=' + obj[key]);
      }
    }
    return args;
  }

  grunt.registerMultiTask('mocha_phantomjs', 'Run client-side mocha test with phantomjs.', function () {
    // Merge options
    var options          = this.options({
          reporter: 'spec',
          urls: [] // Non file urls to test
        }),
        config           = objectAssign({ useColors: true }, options.config),
        phantomPath      = lookup('.bin/phantomjs', true),
        phantomConfig    = objectToArgArray(options.phantomConfig || []),
        mochaPhantomPath = lookup('mocha-phantomjs-core/mocha-phantomjs-core.js'),
        urls             = options.urls.concat(this.filesSrc),
        done             = this.async(),
        errors           = 0,
        output           = options.output || false,
        failWithOutput   = options.failWithOutput || false,
        silent           = options.silent || false,
        reportFailure    = !output || (output && failWithOutput),
        writeStream;

    if (output) {
      grunt.file.mkdir(path.dirname(output));
      writeStream = fs.createWriteStream(output);
    }

    if (grunt.option('color') === false) {
      config.useColors = false;
    }

    async.eachSeries(urls, function (f, next) {
      var args,
          phantomjs;

      args = phantomConfig.concat(flatten([mochaPhantomPath, f, options.reporter, JSON.stringify(config)]));

      phantomjs = grunt.util.spawn({
        cmd: phantomPath,
        args: args
      }, function () {
        next();
      });

      if (!silent) {
        phantomjs.stdout.pipe(process.stdout);
        phantomjs.stderr.pipe(process.stderr);
      }

      // Write output to file
      if (output) {
        phantomjs.stdout.pipe(writeStream, { end: false });
      }

      phantomjs.on('exit', function (code) {
        if (code === 127) {
          grunt.fail.warn('Phantomjs isn\'t installed');
        }
        errors += code;
      });
    }, function () {
      // Fail if errors are reported and we aren't outputing to a file
      if (reportFailure && errors > 0) {
        grunt.fail.warn(errors + ' tests failed');
      }

      if (output) {
        writeStream.end();
      }

      done();
    });
  });
};


Gruntfile.js
/*
 * grunt-mocha-phantomjs
 * https://github.com/jdcataldo/grunt-mocha-phantomjs
 *
 * Copyright (c) 2013 Justin Cataldo
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
  // Project configuration.
  grunt.initConfig({
    eslint: {
      all: ['Gruntfile.js', 'tasks/*.js', 'test/*.js']
    },
    connect: {
      server: {
        options: {
          port: 8000,
          base: '.'
        }
      }
    },
    mocha_phantomjs: {
      no_output: {
        options: {
          reporter: 'dot'
        },
        files: {
          src: ['test/index.html']
        }
      },
      output: {
        options: {
          reporter: 'xunit',
          output: 'results/result.xml'
        },
        files: {
          src: ['test/index.html']
        }
      },
      grep: {
        options: {
          reporter: 'dot',
          config: {
            grep: 'not'
          }
        },
        files: {
          src: ['test/index.html']
        }
      },
      viewport: {
        options: {
          reporter: 'dot',
          config: {
            viewportSize: {
              width: 1024,
              height: 768
            }
          }
        },
        files: {
          src: ['test/index.html']
        }
      },
      phantomConfig: {
        options: {
          phantomConfig: {
            '--debug': true
          }
        },
        files: {
          src: ['test/index.html']
        }
      },
      silent: {
        options: {
          reporter: 'xunit',
          output: 'results/results_silent.xml',
          silent: true
        },
        files: {
          src: ['test/index.html']
        }
      },
      server: {
        options: {
          urls: ['http://localhost:8000/test/index.html'],
          reporter: 'dot'
        }
      },
      coverage: {
        options: {
          urls: ['http://localhost:8000/test-cov/index.html'],
          reporter: 'json-cov',
          output: 'results/coverage.json'
        }
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-connect');
  grunt.loadNpmTasks('grunt-eslint');
  grunt.loadTasks('tasks');

  grunt.registerTask('default', ['eslint', 'connect', 'mocha_phantomjs']);
};
